Documentation/design-docs/hw-intrinsics.md

```diff
@@ -0,0 +1,81 @@
+Implementation of Hardware Intrinsics in CoreCLR
+================================================
+This document describes the implementation of hardware intrinsics in CoreCLR.
+For information about how the intrinsic APIs are designed, proposed and approved,
+see https://github.com/dotnet/designs/blob/master/accepted/platform-intrinsics.md.
+
+In discussing the hardware intrinsics, we refer to the target platform, such as X86 or Arm64, as the "platform" and each set of extensions that are implemented as a unit (e.g. AVX2 on X64 or Simd on Arm64) as an "ISA".
+
+There is a design document for the Arm64 intrinsics: https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/arm64-intrinsics.md. It should be updated to reflect current (and ongoing) progress.
+
+## Overview
+
+The reference assemblies for the hardware intrinsics live in corefx, but all of the implementation is in the coreclr repo:
+
+* The C# implementation lives in coreclr/src/shared/System/Runtime/Intrinsics. These are little more than skeleton methods that usually invoke themselves recursively (see Indirect invocation support below).
+
+## C# Implementation
+
+The hardware intrinsics operate on and produce both primitive types (`int`, `float`, etc.) as well as vector types. The vector types are considered platform-agnostic, though not all platforms define intrinsics on all of these types. They are:
+
+* `Vector64<T>` - A 64-bit vector of type `T`. For example, a `Vector64<int>` would hold two 32-bit integers.
+* `Vector128<T>` - A 128-bit vector of type `T`
+* `Vector256<T>` - A 256-bit vector of type `T`
+
+Note that these are generic types, which distinguishes these from native intrinsic vector types. It also somewhat complicates interop, as the runtime currently doesn't support interop for generic types.
+*** Is there an issue for this?? ***
```

Comment:
I don't think we have a formal issue tracking this, although it has been discussed on several other threads.

CC. @jkotas

Reply:
https://github.com/dotnet/coreclr/issues/1685.

https://github.com/dotnet/coreclr/issues/1685#issuecomment-429361589 is my take on it.